File: hackerrank/problem solving/algorithms/2 - implementation/46 - beautiful triplets/oo/beautiful_triplets_oo.java
// Source: https://www.hackerrank.com/challenges/beautiful-triplets/problem?isFullScreen=true

import java.util.Scanner;
import java.util.List;
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.Collections;

public class Solution {
    private static Scanner _scanner;

    public static void main(String[] args) {
        _scanner = new Scanner(System.in);

        final int size = _readANumber();
        final int beautifulDifference = _readANumber();
        final List<Integer> array = _readAnArray(size);

        final BeautifulTriplets obj = new BeautifulTriplets(array, beautifulDifference);
        System.out.println(obj.getNBeautifulTriplets());
    }

        private static int _readANumber() {
            final int number = _scanner.nextInt();
            return number;
        }

        private static List<Integer> _readAnArray(int size) {
            List<Integer> array = Arrays.asList(new Integer[size]);
            array = array.stream().map(element -> _scanner.nextInt()).collect(Collectors.toList());
            return array;
        }
}

    class BeautifulTriplets {
        private List<Integer> _array;
        private int _beautifulDifference;
        private int _nBeautifulTriplets;

        public BeautifulTriplets(List<Integer> array, int beautifulDifference) {
            _array = array;
            _beautifulDifference = beautifulDifference;
            _findBeautifulTriplets();
        }

            private void _findBeautifulTriplets() {
                _nBeautifulTriplets = 0;

                if (_array.size() > 2)
                    for (int i = 0; i < _array.size() - 2; i++)
                        for (int j = i + 1; j < _array.size() - 1; j++) {
                            final int firstDifference = _array.get(j) - _array.get(i);

                            if (firstDifference < _beautifulDifference)
                                continue;
                            else if (firstDifference > _beautifulDifference)
                                break;

                            for (int k = j + 1; k < _array.size(); k++) {
                                final int secondDifference = _array.get(k) - _array.get(j);

                                if (secondDifference == _beautifulDifference)
                                    _nBeautifulTriplets++;
                                else if (secondDifference > _beautifulDifference)
                                    break;
                            }
                        }
            }

        public int getNBeautifulTriplets() {
            return _nBeautifulTriplets;
        }
    }

File: hackerrank/problem solving/algorithms/2 - implementation/46 - beautiful triplets/oo/beautiful_triplets_oo.cpp
// Source: https://www.hackerrank.com/challenges/beautiful-triplets/problem?isFullScreen=true

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int read_a_number();
vector<int> read_an_array(const int size);

class Beautiful_Triplets
{
private:
    vector<int> array;
    int beautiful_difference;
    int n_beautiful_triplets;

    void find_beautiful_triplets()
    {
        this->n_beautiful_triplets = 0;
        const int array_size {static_cast<int>(this->array.size())};
        if (array_size > 2)
            for (int i {0}; i < array_size - 2; i++)
                for (int j {i + 1}; j < array_size - 1; j++)
                {
                    const int first_difference {array[j] - array[i]};

                    if (first_difference < this->beautiful_difference)
                        continue;
                    else if (first_difference > this->beautiful_difference)
                        break;

                    for (int k {j + 1}; k < array_size; k++)
                    {
                        const int second_difference {array[k] - array[j]};

                        if (second_difference == this->beautiful_difference)
                            this->n_beautiful_triplets++;
                        else if (second_difference > this->beautiful_difference)
                            break;
                    }
                }
    }

public:
    Beautiful_Triplets(const vector<int> array, const int beautiful_difference)
    {
        this->array = {array};
        this->beautiful_difference = {beautiful_difference};
        find_beautiful_triplets();
    }

    int get_n_beautiful_triplets() const
    {
        return n_beautiful_triplets;
    }
};

int main()
{
    const int size {read_a_number()};
    const int beautiful_difference {read_a_number()};
    const vector<int> array {read_an_array(size)};

    const Beautiful_Triplets obj(array, beautiful_difference);
    cout << obj.get_n_beautiful_triplets() << "\n";

    return 0;
}

    int read_a_number()
    {
        int number;
        cin >> number;
        return number;
    }

    vector<int> read_an_array(const int size)
    {
        vector<int> array(size);
        ranges::generate(array, read_a_number);
        return array;
    }
